Alternate, Simplified Async Pull

imuni4fun opened the pull request:
This PR implements an alternative, simplified approach for sync and async image pulling of images. It leverages the previous, synchronous code and behavior as much as possible.

Effectively, only the actual pull activity is affected. Before `v0.7.0`, the pull activity was entirely synchronous and a timeout meant the pull was terminated. Now, the same flow is preserved except that after the existing logic is evaluated and a decision to pull is made, a check is done to see if if the pull is already occurring because it was previously tried and timed out.

Many factors complicate the simple timeout-based cancellation that this asynchronous pull feature addresses. Proper modularity allows for more robust and simple tests of individual logic. The changes to existing code are very minimal. The new behavior is implemented in a modular way in new files (a new package). The behavior is more easily tested because more of the system can be mocked.

For another comparison which does not have to back out as many new changes, see https://github.com/warm-metal/container-image-csi-driver/compare/v0.7.0...imuni4fun:container-image-csi-driver:async-pull-simplified-diff?expand=1
That comparison is missing a few minor tweaks but it illustrates the minimalism of the change to existing code.

Created issue #147 per request in community standup. It shares a little more background on this PR.

imuni4fun commented:
Note that I still need to update logging to use the chosen logger rather than `fmt.Printf()` which I had used in my linked example for the pattern. Logging has also changed since the initial integration that I did pre-`v0.7.0`. Everything other than a few logging lines are ready for evaluation.

imuni4fun commented:
logging is now updated to use klog. verbose logging of internals of the AsyncPuller can be enabled by increasing log verbosity, though i was having trouble trying to pass this through `go test` to get those higher verbosity outputs.

imuni4fun commented:
go test -count=1 -v ./pkg/remoteimageasync/...         
=== RUN   TestPullDuration
pullerMock: starting to pull image nginx:6
pullerMock: starting to pull image nginx:2
pullerMock: starting to pull image nginx:7
pullerMock: starting to pull image nginx:1
pullerMock: starting to pull image nginx:4
pullerMock: starting to pull image nginx:3
pullerMock: starting to pull image nginx:8
pullerMock: finshed pulling image nginx:1
pullerMock: finshed pulling image nginx:2
pullerMock: finshed pulling image nginx:3
pullerMock: finshed pulling image nginx:4
pullerMock: context cancelled
pullerMock: context cancelled
pullerMock: context cancelled
--- PASS: TestPullDuration (5.00s)
=== RUN   TestParallelPull
pullerMock: starting to pull image nginx:2
pullerMock: finshed pulling image nginx:2
--- PASS: TestParallelPull (3.00s)
=== RUN   TestSerialResumedSessions
pullerMock: starting to pull image nginx:A
pullerMock: finshed pulling image nginx:A
--- PASS: TestSerialResumedSessions (6.00s)
=== RUN   TestParallelResumedSessions
pullerMock: starting to pull image nginx:A
pullerMock: finshed pulling image nginx:A
--- PASS: TestParallelResumedSessions (6.00s)
PASS
ok      github.com/warm-metal/container-image-csi-driver/pkg/remoteimageasync   20.607s

mugdha-adhav commented:
Node plugin pod seems to be crashing due to -
```sh
container-image-csi-driver-warm-metal-csi-driver-nodeplugizd67g csi-plugin I0208 11:19:14.556750       1 provider.go:82] Docker config file not found: couldn't find valid .dockercfg after checking in [  /root /]
container-image-csi-driver-warm-metal-csi-driver-nodeplugizd67g csi-plugin panic: runtime error: invalid memory address or nil pointer dereference
container-image-csi-driver-warm-metal-csi-driver-nodeplugizd67g csi-plugin [signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x19f7b2c]
container-image-csi-driver-warm-metal-csi-driver-nodeplugizd67g csi-plugin
container-image-csi-driver-warm-metal-csi-driver-nodeplugizd67g csi-plugin goroutine 44 [running]:
container-image-csi-driver-warm-metal-csi-driver-nodeplugizd67g csi-plugin github.com/warm-metal/container-image-csi-driver/pkg/remoteimage.puller.Pull({{0x0, 0x0}, {0x22a1800, 0x400017c8c0}, {0x2299df8, 0x40007317a0}}, {0x22b3d90, 0x4000a25770})
container-image-csi-driver-warm-metal-csi-driver-nodeplugizd67g csi-plugin 	/go/src/container-image-csi-driver/pkg/remoteimage/pull.go:47 +0x13c
```

Working on fixing this issue.

vadasambar commented on cmd/plugin/main.go:
```diff
@@ -51,8 +51,8 @@ var (
 			"If set to false, secrets will be fetched from the API server on every image pull.")
-	asyncImagePullMount = flag.Bool("async-pull-mount", false,
-		"Whether to pull images asynchronously (helps prevent timeout for larger images)")
+	asyncImagePullTimeout = flag.Duration("async-pull-timeout", 0,
```
👍

vadasambar commented:
I don't think this approach is anymore simple than the previous approach. I would argue it is more difficult to understand overall because it has too many indirections. Maybe this is what we need but I wouldn't call it simplified. It has the advantage of not moving/abstracting/changing the existing `NodePublishVolume` code which is nice.

vadasambar commented on pkg/remoteimageasync/puller.go:
```diff
@@ -0,0 +1,71 @@
+					// 	klog.V(2).Infof("%s.RunPullerLoop(): image not found, pulling %s\n", prefix, ses.image)
+					// 	cri.pullImage(ses.image, ctx2)
+					// }
+					pullErr := ses.puller.Pull(ctxCombined) //NOTE: relying existing tests or history to veirfy behavior, asyncPull just wraps it
```
How is this `Pull` linked to `NodePublishVolume`?

imuni4fun replied:
note that the above statement
> if that pull activity completes during kubelet's backoff then the image would be present already when the next call to NodePublishVolume is made and no pull is needed
assumes `pullPolicy: IfNotPresent` and would likely be untrue for `pullPolicy: Always` in terms of the logic in `node_server.go`. whether or not another pull occurs would be up to containerd since the `remotepuller.Pull()` method WOULD be invoked.

imuni4fun replied:
to say this more simply
```
NodePublishVolume(){
  shouldPull := shouldPullLogic()
  puller:=createPuller()
  if !async
    puller.pull(callerContext) // callerContext is 2m from kubelet
  }else{
    session:=asyncPuller.StartPull(puller,pullerTimeout) // pullerTimeout allows pull to run for longer
    asyncPuller.WaitForPull(session,callerContext) // callerContext is 2m from kubelet
  }
}
```
and in the other go routine you pointed to, the async pull activity calls the same `remoteimage.Pull()` method but calls it with the `pullerTimeout`, not the `callerTimeout`. these are run in separate code paths so that kubelet cancelling after 2m does not cancel the activity. the `session` stays operational until `pullerTimeout` or `container-image-csi` pod is shut down or containerd responds with success or error. these outcomes are captured as the result of the session and the session is marked as completed.

any callers blocked on `remoteimageasync.WaitForPull()` then wake up and observe the result of the activity. remember that they can time out and come back and look up the session that's still ongoing.

imuni4fun commented:
> I don't think this approach is anymore simple than the previous approach. I would argue it is more difficult to understand overall because it has too many indirections. Maybe this is what we need but I wouldn't call it simplified. It has the advantage of not moving/abstracting/changing the existing `NodePublishVolume` code which is nice.

i do believe that this approach is simpler for the following reasons:
- very little changed in `node_server.go` from pre-async
  - i reverted the order of calls back to pre-`v0.7.0` so it looks like more is changed. refer to comparison link in earlier comment.
  - this makes no change to the actual pull behavior. it is still performed by `remoteasync.Pull()` in both async and sync code paths
  - no new logic is introduced regarding determining whether to pull. this decision is made before looking at whether or not we will pull async or sync
  - mounting is still synchronous. less states to manage than in `v0.7.0`
- interfaces are cleaner and more testable
  - a new `pullerMock` object was used for unit testing and verifying continuation of pulls
  - mocking allowed clean unit test of new logic without integration with containerd
  - unit test now exercises relevant async continuation logic
- generally cleaner multi-threading implementation
  - prefer less mutexes, no polling
  - you almost always want to lock the smallest segment of code possible which doesn't allow for race conditions
  - channels are preferred over mutexes
  - sleeping is preferred over polling
  - this structure best represents the actual problem: pull session is really the fundamentally most pure representation of what is happening that i can think of. then want to reduce edge cases and states as much as possible. drive simplicity of states and data structures. coming from deep history with multi-threading across many domains, the problems that i am familiar with are best solved via this "session" approach.

imuni4fun commented on pkg/metrics/metrics.go:
```diff
@@ -11,27 +11,26 @@ import (
-		Buckets:   []float64{0, 1, 5, 10, 15, 30, 60, 120, 180},
+		Buckets:   []float64{1, 5, 10, 15, 30, 60, 120, 300, 600, 900},
 	},
-	[]string{"operation_type"},
```
there is no sync vs async in this implementation... they actual pull is identical. this PR changes the logic surrounding the previously sync-only puller. whether sync or async, the original `remoteimage.Pull()` is used relatively unchanged.

imuni4fun commented on pkg/metrics/metrics.go:
```diff
@@ -11,27 +11,26 @@ import (
 		Help:      "The time it took to mount an image",
-		Buckets:   []float64{0, 1, 5, 10, 15, 30, 60, 120, 180},
 	},
-	[]string{"operation_type"},
```
same as above. additionally, reporting based on image and result will be more useful. go routine is launched to delete the value after 1m so we wont have values building up. included this change since some of the specific metrics had to change anyway with this implementation change.

imuni4fun commented on pkg/remoteimage/pull.go:
```diff
@@ -40,14 +43,31 @@ func (p puller) ImageSize(ctx context.Context) int {
 }
 
 func (p puller) Pull(ctx context.Context) (err error) {
+	startTime := time.Now()
```
this is the only real change to the actual pull code... how metrics are emitted in a centralized, consistent way regardless of async/sync

imuni4fun commented on pkg/remoteimage/pull.go:
```diff
@@ -40,14 +43,31 @@ func (p puller) ImageSize(ctx context.Context) int {
+			// pushgateway may remove the need for this. https://prometheus.io/docs/practices/pushing/
+			time.Sleep(1 * time.Minute)
+			metrics.ImagePullTime.DeleteLabelValues(p.image.String(), metrics.BoolToString(err != nil))
+		}()
```
note the 1m cleanup logic spawned in separate go routine. this prevents pile-up of metrics for prometheus to scrape.

imuni4fun commented on test/integration/metrics-manifests/rendered-test.yaml:
```diff
@@ -1,25 +0,0 @@
-apiVersion: batch/v1
```
this file should not be here. added to a `.gitignore` file for another PR for issue #143 which captured a few developer experience improvements (IMHO)

imuni4fun commented:
@mugdha-adhav could you re-trigger the integration tests?

imuni4fun commented:
@mugdha-adhav it looks like I cannot request reviews from specific people. I will at them here instead.

@kitt1987 @MaxFedotov @mbtamuli could you take a peek at this PR? thanks!

imuni4fun commented:
decided to re-introduce the size extraction and logging since it was present in the reverted files which were removed. added more error handling because there was none and i experienced a panic in my local dev env. added metric export as a bonus since it was very little work and followed same pattern as image pull time (individual).

imuni4fun commented:
@mugdha-adhav could you kick integration tests again when you get the chance? thanks!

imuni4fun commented on pkg/remoteimage/pull.go:
```diff
@@ -31,23 +35,76 @@ type puller struct {
 // Returns the compressed size of the image that was pulled in bytes
 // see https://github.com/containerd/containerd/issues/9261
-func (p puller) ImageSize(ctx context.Context) int {
+func (p puller) ImageSize(ctx context.Context) (size int, err error) {
```
added error checking to this. experienced a panic loop in my local environment.

imuni4fun commented on pkg/remoteimage/pull.go:
```diff
@@ -31,23 +35,76 @@ type puller struct {
+			metrics.OperationErrorsCount.WithLabelValues("pull-error").Inc()
+		}
+		go func() {
+			//TODO: this is a hack to ensure data is cleared in a reasonable time frame (after scrape) and does not build up.
```
this hack is mentioned above. spawns go routine to sleep. could instead add to a channel with a timestamp and an anonymous func. would have to guard against deadlock scenarios.

imuni4fun commented on pkg/remoteimage/pull.go:
```diff
@@ -31,23 +35,76 @@ type puller struct {
+				klog.Infof("remoteimage.Pull(): pulled %s with size of %d bytes", p.image.String(), size)
+				metrics.ImagePullSizeBytes.WithLabelValues(p.image.String()).Set(float64(size))
+				go func() {
+					//TODO: this is a hack to ensure data is cleared in a reasonable time frame (after scrape) and does not build up.
```
see similar hack comment a few lines earlier

imuni4fun commented on pkg/remoteimageasync/synchronizer.go:
```diff
@@ -0,0 +1,119 @@
+// NOTE: all sessions that are successfully submitted to sessionsChan must be submitted to completedEvents
+func (s synchronizer) RunCompletionsLoop() {
+	go func() {
+		klog.V(2).Infof("%s.RunCompletionsLoop(): starting", prefix)
```
generally simplified the states in this file since initial implementation

imuni4fun commented on pkg/remoteimageasync/puller.go:
```diff
@@ -0,0 +1,70 @@
+		<-ctx.Done()
+		close(sessionChan) // only close this once
+	}()
+	go func() {
```
generally simplified the states in this file since initial implementation

mugdha-adhav commented:
@imuni4fun currently we are not testing the async mode in our CI tests. Could you please add a test which enables the async feature and then runs integration tests for it?

imuni4fun commented on cmd/plugin/node_server.go:
```diff
@@ -31,36 +30,36 @@ const (
+		ns.asyncImagePuller = remoteimageasync.StartAsyncPuller(context.TODO(), 100, 20)
+	} else {
+		klog.Info("Starting node server in Sync mode")
+		ns.asyncImagePullTimeout = 0 // set to default value
```
because the comparison is to 30s, not equality to zero. this ensures that it is either async and >=30s or 0s.

imuni4fun replied:
this could be avoided if we used a nil check for the async puller or wrapped any checks in a helper method. this avoids the case where 10s would be not be considered async here but may be considered async later due to different criteria being used.

imuni4fun replied:
for clarity, the `node_server.go` code currently DOES use nil check to see if running in async mode but someone could also have read the Cobra parameter description and tried to use the async timeout DURATION itself. this could be error prone over time. instead, i'm forcing consistency here.

vadasambar commented on pkg/remoteimageasync/puller.go:
```diff
@@ -0,0 +1,70 @@
+				select {
+				case <-ctx.Done(): // shutting down
+					ses.isComplete = false
+					ses.isTimedOut = false
```
Do we need `isComplete` and `isTimedOut` ?

imuni4fun replied:
per our discussion earlier today, i have now removed `isComplete`. i left `isTimeout` since i felt that trying to pack the error type into the error message in a rigid way was becoming bloated and made the code harder to read and hinder debugging as the structure of logging would be changed.

vadasambar commented on pkg/remoteimageasync/synchronizer.go:
```diff
@@ -0,0 +1,119 @@
+}
+
+// NOTE: all sessions that are successfully submitted to sessionsChan must be submitted to completedEvents
+func (s synchronizer) RunCompletionsLoop() {
```
Can we clean the image from the session immediately after the image is pulled? Channel blocking here worries me a little.  Doing it immediately after the pull sounds simpler.

imuni4fun replied:
agree. we discussed this today in our zoom. this has been replaced with an anonymous func to be run in same go routine at completion.

imuni4fun commented:
@mugdha-adhav could you kick integration tests again? thanks

imuni4fun commented on cmd/plugin/node_server.go:
```diff
@@ -31,36 +30,36 @@ const (
 }
 
 func (n NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (resp *csi.NodePublishVolumeResponse, err error) {
-	valuesLogger := klog.LoggerWithValues(klog.NewKlogr(), "pod-name", req.VolumeContext["pod-name"], "namespace", req.VolumeContext["namespace"], "uid", req.VolumeContext["uid"], "request-id", uuid.NewString())
```
this is the only place that it was used after this PR. it didn't relate to anything else and wasn't stored for association with future log messages, etc. didn't add any value in current form.

imuni4fun commented on cmd/plugin/node_server.go:
```diff
@@ -122,56 +121,59 @@ func (n NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishV
-		err = status.Errorf(codes.Aborted, "unable to pull image %q: %s", image, e)
+	keyring, err := n.secretStore.GetDockerKeyring(ctx, req.Secrets)
+	if err != nil {
+		err = status.Errorf(codes.Aborted, "unable to fetch keyring: %s", err)
```
i'm not crazy about this syntax either. this DOES return it. this is using named return variables. whatever is in the `err` value at the top level function scope when it exits is the value returned.

we can either `return myResponse,myError` or just `return` which is equal to `return resp,err` (i think this is at top level function scope, not immediate scope if `err` was redefined within a smaller scoped block)

imuni4fun commented on pkg/remoteimageasync/puller.go:
```diff
@@ -0,0 +1,70 @@
+			ses, ok := <-sessionChan // ctx not observed for shut down, this sleep breaks when sessionChan is closed
+			if !ok {                 // sessionChan closed, shut down loop
+				wg.Wait()            // wait for running goroutines (which observe cancellation)
+				close(completedChan) // only close this once, after completions submitted (wg.Wait() returns)
```
no, sessionChan must be closed by the sender. `RunPullerLoop()` is the receiving activity. this part of Go channels design is a mess IMHO. it should allow checking for closed when writing just like when reading by returning second optional result. instead, it panics. the theory is that a producer should not close a channel without coordinating with other producers (complexity). i would prefer if that coordination was as i just described instead of needing more synchronization primitives and one-off design.

imuni4fun replied:
writing to the channel already has to synchronize so it should offer the "is closed" check for free IMHO. we could still implement our own coordination if we could not start an activity unless we knew we could submit that to the channel... but now we need to implement that for all cases. the example here was one where it was a bit complex to determine how to close the channels because there could always be another `NodePublishVolume()` call to come it after `sessionChan` was closed.

imuni4fun commented on pkg/remoteimageasync/synchronizer.go:
```diff
@@ -0,0 +1,119 @@
+		completedChan,
+	)
+	async.RunCompletionsLoop()
+	RunPullerLoop(ctx, sessionChan, completedChan)
```
i created the interface for external callers. this is an internal detail IMHO.

imuni4fun commented on pkg/remoteimageasync/synchronizer.go:
```diff
@@ -0,0 +1,119 @@
+	}
+}
+
+func (s synchronizer) StartPull(image string, puller remoteimage.Puller, asyncPullTimeout time.Duration) (ses *PullSession, err error) {
```
the `asyncPullTimeout` is passed into the `AsyncPuller.StartPull()` so that it can be varied by from image to image. this is in case we decide to allow that to be changed in the volume definition. this might not be desirable or possible so this could be removed as you say. this implementation could accept the same value every time or a different value per image so i exposed it that way. the other behavior was left to the next level up (node_server.go).

imuni4fun replied:
you're correct, we could probably rely on extracting the image name from `puller.Image()`. i didn't take a dependency on that here.

vadasambar commented on pkg/remoteimageasync/synchronizer_test.go:
```diff
@@ -0,0 +1,281 @@
+	"github.com/stretchr/testify/assert"
+)
+
+type chanTestStruct struct {
```
Can we use the actual session object here?

vadasambar commented on pkg/remoteimageasync/synchronizer_test.go:
```diff
@@ -0,0 +1,281 @@
+}
+
+// demonstrates session channel structure's pass-by-reference is appropriate
+func TestChannelStructContent(t *testing.T) {
```
Same comment as above. If these tests are not directly or indirectly related by code to synchronizer, I think we should put it in a separate file.

vadasambar commented:
Overall LGTM provided we 
* address https://github.com/warm-metal/container-image-csi-driver/pull/137#discussion_r1500686922 
* address https://github.com/warm-metal/container-image-csi-driver/pull/137#discussion_r1500694509
* add tests for `NodePublishVolume` (in node_server_test.go)

imuni4fun commented:
summary of discussion with @vadasambar:

I agree that it would be good to build out additional regression tests at higher layers of the onion... in this case at the call to `PublishNodeVolume()` with mocked containerd being called over GRPC.

I believe this additional testing does not exercise additional code paths beyond one case where there is interplay between caller timeout (kubelet) and actual pull time when pull policy is always. In this case, an image pull that takes too long to pull and exceeds the 2m caller timeout but ALSO completes the pull BEFORE the next call to continue (completes during backoff waiting period) will start a new pull on next call. The image in this scenario will never succeed. Because it would never succeed previously, and just timeout and cancel, I consider this the same behavior.

The above scenario tests the pre-existing pull-always logic in a new context where now many more scenarios CAN succeed. Anything that cannot succeed after the change would also have not succeeded prior.

I am happy to grow the regression tests but don't think adding these new layers should block this feature.

imuni4fun commented:
@vadasambar @mugdha-adhav could you review? thx

imuni4fun commented:
@mugdha-adhav created meeting to wrap up tomorrow

could you kick off integration tests again? it looks like it might need another run but it’s not clear from my end

last commit was just comments. previous commit had passing integration tests

imuni4fun commented:
> Also why did we remove the metrics test?

I didn’t remove the test. I removed the “rendered” version which is rendered from https://github.com/imuni4fun/container-image-csi-driver/blob/4c6d7160fa72408225477aad4b061765753095e8/test/integration/metrics-manifests/test.yaml#L1

rendered test changes each time and should be ignored. i added it to .gitignore file in different PR/issue.
:point_up: https://github.com/warm-metal/container-image-csi-driver/pull/148/files#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947

mugdha-adhav commented on charts/warm-metal-csi-driver/templates/nodeplugin.yaml:
```diff
@@ -63,8 +63,8 @@ spec:
-            {{- if .Values.enableAsyncPullMount }}
-            - --async-pull-mount=true
+            {{- if .Values.enableAsyncPull }}
+            - --async-pull-timeout=10m
```
After discussing with @imuni4fun, we decided that we want to keep the `enableAsyncPull` and adding default timeout value to be passed in from values.yaml. Something like `--async-pull-timeout={{ .Values.asyncPullTimeout }}`

mugdha-adhav commented on pkg/remoteimage/pull.go:
```diff
@@ -29,25 +35,86 @@ type puller struct {
+	repo := p.ImageWithoutTag()
+	imageSpec := &cri.ImageSpec{Image: p.ImageWithTag()}
 	creds, withCredentials := p.keyring.Lookup(repo)
+	// klog.V(2).Infof("remoteimage.Pull(): len(creds)=%d, withCreds=%t", len(creds), withCredentials)
```
We could remove this comment if we don't need it.

woehrl01 commented on pkg/remoteimage/pull.go:
```diff
@@ -68,11 +135,14 @@ func (p puller) Pull(ctx context.Context) (err error) {
 		})
 
 		if err == nil {
+			klog.V(2).Info("remoteimage.Pull(with creds): pulling %s completed with err==nil", p.image.String())
```
this is missing a `Infof` instead of Info

imuni4fun replied:
good catch!

imuni4fun commented:
@mugdha-adhav pushed changes called out in review plus from @woehrl01 above. also one small replacement of docker.Named.String() with Puller.ImageWithTag()

ephemeral-volumet-set test passed in local kind cluster
